### PR TITLE
Scripting: Per-context script cache, default off

### PR DESCRIPTION
**Pull request by stu-elastic**
* Adds per context settings:
  `script.context.${CONTEXT}.cache_max_size` ~
  `script.cache.max_size`

  `script.context.${CONTEXT}.cache_expire` ~
  `script.cache.expire`

  `script.context.${CONTEXT}.max_compilations_rate` ~
  `script.max_compilations_rate`

* Context cache is used if:
  `script.max_compilations_rate=use-context`.  This
  value is dynamically updatable, so users can
  switch back to the general cache if desired.

* Settings for context caches take the first value 
  that applies:
  1) Context specific settings if set, eg
     `script.context.ingest.cache_max_size`
  2) Correlated general setting is set to the non-default 
     value, eg `script.cache.max_size`
  3) Context default

The reason for 2's inclusion is to allow an easy
transition for users who've customized their general
cache settings.

Using the general cache settings for the context caches
results in higher effective settings, since they are 
multiplied across the number of contexts.  So a general
cache max size of 200 will become 200 * # of contexts.
However, this behavior it will avoid users snapping to a
value that is too low for them.

Refs: #50152

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (jdconrad) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: After looking through the code more carefully, we can't do this because settings is never actually updated for dynamic settings. This would only determine if this setting exists at start-up. So if a user were to dynamically update this setting we would be using old information.

**Review by jdconrad — CHANGES_REQUESTED** [review]
I spent more time thinking about how we use `SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED` as a sentinel value for determining whether we have overlapping deprecated settings versus new context settings, and I think we need to come up with something different :/ See my comment at those specific lines.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: What do you think about changing this to:

`SCRIPT_MAX_COMPILATIONS_RATE_DEPRECATED.getDefault(Settings.EMPTY).equals(s)`

That will detect non-changes.

Comment (rjernst) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: That would only detect if the value was different than the default. The user could explicitly set the default through a dynamic update?

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: I don't think we've used this naming pattern of suffixing with _DEPRECATED before. If we are going to change the object name, perhaps instead we could refer to this as the "general" version since that is the terminology you used elsewhere? eg SCRIPT_GENERAL_CACHE_SIZE_SETTING. We also need the Property.Deprecated on these.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: call caps case?

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: Why are we not just using the exists method?

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: This looks like the case for cross setting validation. Whether the general cache should be created should strictly be based on any general cache settings existing. Validating both sets of settings don't exist at the same time is a separate issue, and can be done by the settings validators themselves.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Yes, the user can set that and we will not reject it.  I address this in the comment on [line 397](https://github.com/elastic/elasticsearch/pull/52855/files#diff-e7f789c9bbf345278cfb259a0b9ac87fR397):
```
// This validator is run on every settings update.
// Here we try to detect if it's a spurious update to the default.
// This _will_ allow through bad updates to the default.
```

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: I think we need to consider changing the signature here to allow the old value to be passed through. We have it in the update consumer within Setting. Also, using the cross setting validation should allow simpler validation here I think. See the other validate methods that can be implemented on Setting.Validator.

**Review by rjernst — COMMENTED** [review]
I have a few comments

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: > We also need the Property.Deprecated on these.

That will be added in a future change, see [step 5](https://github.com/elastic/elasticsearch/issues/50152#issuecomment-587927041)

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (jdconrad) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: @stu-elastic Thanks for walking me through this. I believe you are correct that if we remove the exists check, this will behave as expected.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: I can pull these out.  I figured I'd just keep the same shape since it doesn't work for Affix settings.
```java
ScriptService.SCRIPT_CACHE_EXPIRE_SETTING.exists(
            Settings.builder()
                .put("script.context.baz.cache_expire", TimeValue.timeValueMillis(1000))
                .build()
        )
```
is false.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: > Validating both sets of settings don't exist at the same time is a separate issue, and can be done by the settings validators themselves.

Let's discuss how this can be done.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Changed.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Changed to `SCRIPT_GENERAL_`, leaving the deprecated property addition to a future change.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Using exists now.

**Comment by stu-elastic** [comment]
Blocking this change on fixing affix settings in https://github.com/elastic/elasticsearch/issues/52933

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (jdconrad) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: `ZERO` is pretty vague as a name here, but I see it matches TimeValue. Maybe this could be `SCRIPT_COMPILATION_RATE_ZERO` or something like that to match `SCRIPT_COMPILATION_RATE`.

**Review by jdconrad — APPROVED** [review]
LGTM. Thanks for all the work here. One very minor comment.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Changed.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Now we have a special `use-context` value.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: As mentioned, deprecation will happen in a future change.  There's a bunch of tests, etc that need to change in tandem with that.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: We're no longer worrying about cross validation.  Now we allow any combo of settings to be set and flip between context and general caches using `script.max_compilations_rate=use-context`.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: We've taken a step back and are changing the semantics here as mentioned above.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/script/ScriptService.java]: One problem that seems to make this more complicated is having individual update consumers. There is an existing method we can use that takes a list of settings, and passes all the values when any are updated. I think using this pattern will make the code simpler (less locking and special cases). The pattern I'm imagining is something like an update consumer for the following scenarios:

- general settings -> this changes the general cache, and handles flipping between general and per context
- per context + general settings -> the general settings along with concrete settings for each context. if the general cache is used, this would result in an error if any of the per context settings exist. otherwise the ScriptCache for that context would be looked up. All 3 setting values should then be passed into a single update method on ScriptCache, so it can be rebuilt once whether one or all 3 settings change (or the fallbacks for size/expire changed).

**Review by rjernst — CHANGES_REQUESTED** [review]
This is looking good. I like the new semantics. However I think we can further simplify.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Future looking comment based on implementation.  Will remove.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: They were never dynamically updatable before, why add them for deprecated settings?

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Appended.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Mimeographed.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Asserted.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: From direct convo, not necessary to add.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Moved.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Moved.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Added fallback.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Added fallback.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Done.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Using fallback settings, thanks for pointing it out.

File: server/src/main/java/org/elasticsearch/script/ScriptService.java
Comment (stu-elastic) [reply server/src/main/java/org/elasticsearch/script/ScriptService.java]: Using fallback settings.

**Review by jdconrad — APPROVED** [review]
LGTM after all these changes. Thanks for cleaning this up so much.

**Review by rjernst — APPROVED** [review]
LGTM

**Comment by stu-elastic** [comment]
master: 070ea7e
7.x:  cdbee32

**Pull request merged**
